moveos/moveos-store/src/state_store/statedb.rs

```diff
@@ -297,6 +306,34 @@ impl StateDBStore {
         }
     }
 
+    // rebuild statedb via StateSet from dump
+    pub fn apply(&self, state_set: StateSet) -> Result<H256> {
+        let mut state_root = H256::zero();
+        for (k, v) in state_set.state_sets.into_iter() {
+            if k == state_resolver::GLOBAL_OBJECT_STORAGE_HANDLE {
+                state_root = self.global_table.puts(v)?
+            } else {
+                // must force create table
+                let (_, table_store) = self.create_table(k)?;
+                state_root = table_store.puts(v)?
+            }
+        }
+        Ok(state_root)
+    }
+
+    // pub fn dump_iter(
+    //     &self,
+    //     handle: &ObjectID,
+    // ) -> Result<Option<SMTIterator<Vec<u8>, State, NodeDBStore>>> {
+    //     if handle == &state_resolver::GLOBAL_OBJECT_STORAGE_HANDLE {
+    //         self.global_table.iter().map(|v| Some(v))
+    //     } else {
+    //         self.get_as_table(*handle)
+    //             .and_then(|res| res.map_or(Ok(None), |(_, table)| table.iter().map(|v| Some(v))))
```

Comment:
There is an problem with lifetime constraints, due to `SMTIerator<'a, K, V, R>` causing compilation failure:

```
   --> moveos/moveos-store/src/state_store/statedb.rs:332:67
    |
332 |                 .and_then(|res| res.map_or(Ok(None), |(_, table)| table.iter().map(|v| Some(v))))
    |                                                                   ------------^^^^^^^^^^^^^^^^^
    |                                                                   |
    |                                                                   returns a value referencing data owned by the current function
    |                                                                   `table` is borrowed here
```

or rewritten as something like
```
    pub fn dump_iter(
        &self,
        handle: &ObjectID,
    ) -> Result<Option<SMTIterator<Vec<u8>, State, dyn NodeStore + 'static>>> {
...
}
```

but it will also compile failure, error stack:
```
error[E0277]: the size for values of type `(dyn NodeStore + 'static)` cannot be known at compilation time
   --> moveos/moveos-store/src/state_store/statedb.rs:327:10
    |
327 |     ) -> Result<Option<SMTIterator<Vec<u8>, State, dyn NodeStore + 'static>>> {
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `(dyn NodeStore + 'static)`
```

Any suggestions for better writing? @jolestar

Reply:
I will give it a try.